### ISSUE TITLE
forceignore doesn't retrieve correctly when using an opt-in style

### ISSUE THREAD

**Issue by aidan-harding** [forcedotcom/cli#2404]
### Summary

The `.forceignore` file is supposed to work like a `.gitignore`, so it should be possible to write an opt-in sort of file like we can in git.

For example, this file would only allow classes:

```
# First, ignore everything
*

# Now, allow anything that's a directory
!*/

# Then allow anything in the classes directory
!**/classes/**
```

However, if I use a file like this, retrieve does not work as expected. A new remote class is listed by retrieve preview but won't be retrieved during an actual retrieve.

### Steps To Reproduce

No repo required. Just create an empty project and scratch org using the `.forceignore` file from above.

We can confirm that creating and deploying a class works:

```
sf apex generate class --name MyClass --output-dir force-app/main/default/classes
sf project deploy start 
```

As expected, MyClass is created in the scratch org.

Now, open the scratch org and use Dev Console to create a class called RemoteClass. Going back to the CLI:

```
sf project retrieve preview
No conflicts found.

No files will be deleted.

Will Retrieve [1] files.
 Type      Fullname    Path 
 ───────── ─────────── ──── 
 ApexClass RemoteClass      

Ignored [1] files. These files won't retrieve because they're ignored by your .forceignore file.
 Type    Fullname Path 
 ─────── ──────── ──── 
 Profile Admin         
```

This is still as expected, so do the retrieve:

```
sf project retrieve start
```

### Expected result
I would expect RemoteClass to be retrieved, as reported in the preview.

### Actual result
Nothing is retrieved:

```
Retrieving v58.0 metadata from test-3vnmdnapvkxv@example.com using the v58.0 SOAP API
Preparing retrieve request... Succeeded
 ›   Warning: Nothing retrieved
```

### System Information
Using zsh shell on mac.

```json
{
  "cliVersion": "@salesforce/cli/2.3.8",
  "architecture": "darwin-arm64",
  "nodeVersion": "node-v18.15.0",
  "osVersion": "Darwin 22.5.0",
  "shell": "zsh",
  "rootPath": "/usr/local/lib/sf",
  "pluginVersions": [
    "@oclif/plugin-autocomplete 2.3.3 (core)",
    "@oclif/plugin-commands 2.2.21 (core)",
    "@oclif/plugin-help 5.2.15 (core)",
    "@oclif/plugin-not-found 2.3.34 (core)",
    "@oclif/plugin-plugins 3.1.8 (core)",
    "@oclif/plugin-search 0.0.21 (core)",
    "@oclif/plugin-update 3.1.28 (core)",
    "@oclif/plugin-version 1.3.8 (core)",
    "@oclif/plugin-warn-if-update-available 2.0.46 (core)",
    "@oclif/plugin-which 2.2.29 (core)",
    "@salesforce/cli 2.3.8 (core)",
    "apex 2.3.9 (core)",
    "auth 2.8.10 (core)",
    "data 2.5.4 (core)",
    "deploy-retrieve 1.16.5 (core)",
    "info 2.6.35 (core)",
    "limits 2.3.27 (core)",
    "login 1.2.23 (core)",
    "org 2.9.28 (core)",
    "schema 2.3.21 (core)",
    "settings 1.4.21 (core)",
    "sobject 0.1.42 (core)",
    "source 2.10.28 (core)",
    "telemetry 2.2.3 (core)",
    "templates 55.5.7 (core)",
    "trust 2.5.1 (core)",
    "user 2.3.27 (core)"
  ]
}
```

**Comment by cristiand391** [comment]
this was fixed in sf v2.10.2:
https://github.com/forcedotcom/cli/tree/main/releasenotes/#2102-sept-27-2023-stable

@aidan-harding you mentioned in cli-friends this also happened in devops center, their PM is aware of this and they should update SDR to include the fix and bump CLI if they are using it (or maybe they already did, pinged them a few weeks ago 😄 )

**Comment by aidan-harding** [comment]
Thank you @cristiand391!
